### PR TITLE
Set up back end of searching for programming expressions

### PR DESCRIPTION
**Pull request by dmcavoy**
Set up the back end that searches through programming expressions based on programming_environment_id plus programming_expression name and category.  This work is largely based on how vocab and resources are set up for editing in lessons.

## Links

- [Technical Design Doc Code Documentation](https://docs.google.com/document/d/1dIETQq4yYRKyHwJWf7V38Y1FqR7JBVtVKzAFrQZYcKg/edit#)
- jira ticket: [PLAT-697](https://codedotorg.atlassian.net/browse/PLAT-697)

## Testing story

Added a basic set of unit tests to check that searching is working.

## Follow-up work

- Setting up the UI in Lesson Edit that uses this search to add programming expressions to lessons (https://github.com/code-dot-org/code-dot-org/pull/39382/files)

## PR Checklist:

- [x] Tests provide adequate coverage
- [x] Privacy and Security impacts have been assessed
- [x] Code is well-commented
- [x] New features are translatable or updates will not break translations
- [x] Relevant documentation has been added or updated
- [x] User impact is well-understood and desirable
- [x] Pull Request is labeled appropriately
- [x] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: dashboard/lib/programming_expression_autocomplete.rb [hunk 1]
```diff
@@ -0,0 +1,13 @@
+class ProgrammingExpressionAutocomplete < AutocompleteHelper
+  def self.get_search_matches(query, limit, programming_environment_id)
+    limit = format_limit(limit)
+
+    rows = ProgrammingExpression.limit(limit)
+    rows = rows.where(programming_environment_id: programming_environment_id) if programming_environment_id
+    return [] if query.length < MIN_WORD_LENGTH
```
Comment (Hamms) [inline dashboard/lib/programming_expression_autocomplete.rb]: any reason not to do this first, to save a query?

File: dashboard/lib/programming_expression_autocomplete.rb [hunk 2]
```diff
@@ -0,0 +1,13 @@
+class ProgrammingExpressionAutocomplete < AutocompleteHelper
```
Comment (Hamms) [inline dashboard/lib/programming_expression_autocomplete.rb:1]: I believe this should be in `dashboard/lib/queries`; see https://github.com/code-dot-org/code-dot-org/tree/staging/dashboard/lib/queries for an explanation of Query Objects

**Review by Hamms — COMMENTED** [review]
Given that this back end is providing an API, I'd suggest we move the controller for this functionality into `api/v1`, and namespace the route appropriately.

I _love_ that the core functionality here has been split off into a PORO! Tests look good, too. Got just a couple reorganization suggestions in addition to the one above.

File: dashboard/lib/programming_expression_autocomplete.rb (hunk 2 above)
Comment (bethanyaconnor) [reply dashboard/lib/programming_expression_autocomplete.rb:1]: `dashboard/lib` is where all the other autocomplete classes live, including the superclass. Not against moving them but I think that's out of the scope of this PR

Comment (Hamms) [reply dashboard/lib/programming_expression_autocomplete.rb:1]: ah, fair. Let's move em all together, then, but in a follow-up PR

Comment (dmcavoy) [reply dashboard/lib/programming_expression_autocomplete.rb:1]: https://codedotorg.atlassian.net/browse/PLAT-842

**Comment by davidsbailey** [comment]
> Given that this back end is providing an API, I'd suggest we move the controller for this functionality into `api/v1`, and namespace the route appropriately.

@Hamms I agree that this is the pattern we mostly follow, but I am not clear on the reason -- AFAIK we only had api/v1 so that it could be shared between dashboard and pegasus. I don't see anything in Rails docs or other best practices about using a separate url namespace for json APIs. Am I missing a reason that we put JSON APIs in api/v1?

Not suggesting we immediately change this, since I'd rather have one system than be half way between two... but wanting to understand what we think the ideal state would be.

**Review by bethanyaconnor — APPROVED** [review]
Looks good!

**Comment by Hamms** [comment]
I'd argue that namespacing APIs is a general web design best practice, not even a rails-specific thing. That being said, I do believe that rails does generally conform to this practice; I'm not sure where the docs would talk about this, but it's certainly the way that things like `rails new` work when you ask it to scaffold you an API.

My understanding is that we originally built API functionality in order to facilitate sharing between dashboard and pegasus, and that we namespaced it because that's just what you do when building an API.

**Pull request merged**